### ISSUE TITLE
Improve heredoc documentation around method arguments 

### ISSUE THREAD

**Issue by Daniel-Worrall** [crystal-lang/crystal#10268]
E: Original issue is invalid. This is now about improving the HEREDOC docs.

The use case is a multiline interpolated text where indents don't affect the whitespace at the beginning of each line to be used directly in a method call.

This isn't possible due to a conflict where the heredoc delimiter needing to be the only thing on the final line, but the comma to separate parameters also needs to be on the same line as the final line of the parameter. This also doesn't work if we leave a space between the delimiter and the comma.

The current alternatives are by either assigning the text outside the call, a begin/end pair, or %/%q.

This *does* work when it is the last parameter since the closing bracket can be on a newline, whereas the comma cannot.

## Reproducible code

[Carc link](https://carc.in/#/r/a9rl)
```cr
def foo(foo1 = nil, foo2 = nil)
  foo1
  foo2
end
 
foo(<<-HEREDOC
  words
  HEREDOC,
  "test",
)
```

```
syntax error in eval:10
Error: Unterminated heredoc: can't find "HEREDOC" anywhere before the end of file
```

## Current Undesired Solutions

Assigning outside the call.
```cr
bar = <<-HEREDOC
  words
  HEREDOC
foo(
  bar,
  "test",
)
```

begin/end is a bit messy, and it shouldn't be necessary.
```cr
foo(begin<<-HEREDOC
  words
  HEREDOC
  end,
  "test",
)
```

%/%q doesn't let us use whitespace to indent the text without also including it in the text. It also parses lines differently
```cr
foo(%(line1
line2
line3),
  "test",
)
```

Tested in both latest 0.35.1 and nightly as of 18th Jan

**Comment by tipsypastels** [comment]
Can't do to this? This works.

```crystal
foo(<<-HEREDOC, "test")
  words
  HEREDOC
```

[Carc](https://carc.in/#/r/a9s3)

**Comment by Daniel-Worrall** [comment]
That works, but it really feels like it shouldn't. Also, this would only support 1 heredoc in the method.

It is also limited to having the other parameters on a single line, which I do not want for my example since there are many parameters for the method call in my use-case

**Comment by asterite** [comment]
Can you show us your real code?

**Comment by Daniel-Worrall** [comment]
```cr
Discord::Embed.new(
    title: title,
    description: desc,
    timestamp: event.start.time,
    thumbnail: lookup_image(event.summary),
    colour: lookup_colour(event)
  )
  ```
  
Here is a snippet of my use-case. The description is the HEREDOC with interpolation based on variables. This happens inside some iteration nesting too. I am choosing to assign to desc beforehand in the meantime.

**Comment by asterite** [comment]
So something like this?

```crystal
Discord::Embed.new(
  title: <<-HEREDOC,
  One
  Two
  HEREDOC
  description: <<-HEREDOC,
  Three
  Four
  HEREDOC
  timestamp: event.start.time,
  thumbnail: lookup_image(event.summary),
  colour: lookup_colour(event)
)
```

Basically heredoc is letters, numbers or underscore. Any whitespace or symbol stops the heredoc start but then regular code can follow. Next lines upto the heredoc identifier are the heredoc body. So if you need a comma you need to put it in the first line.

Other examples:

```crystal
foo(a, b, <<-HEREDOC.upcase, c)
  FOO BAR
HEREDOC

bar(a, <<-HEREDOC + " world!")
  Hello
HEREDOC
```

By the way, this is exactly how it works in Ruby too.

**Comment by Daniel-Worrall** [comment]
Ah! Thank you, I was not aware of needing to put the comma in the first line. It seems very confusing/counter-intuitive. Maybe the reference could be expanded to include an example like this

**Comment by asterite** [comment]
I agree, the reference doesn't make this clear: https://crystal-lang.org/reference/syntax_and_semantics/literals/string.html

@oprypin By the way, can the doc nav table be expanded inline? I find it a bit hard to see at a glance what's the table of contents. Previously I would expand "Syntax and semantics", then "Literals", then "String". In fact it was expanded beforehand. With the new approach I feel pretty restricted.

**Comment by oprypin** [comment]
@asterite there are [a few](https://squidfunk.github.io/mkdocs-material/setup/setting-up-navigation/#navigation-expansion) [things](https://squidfunk.github.io/mkdocs-material/setup/setting-up-navigation/#navigation-sections) that could be done. But most of them will be blocked on https://github.com/squidfunk/mkdocs-material/issues/2155 because the nav is too huge. You can thumb up that issue :D

**Comment by straight-shoota** [comment]
Fixed by crystal-lang/crystal-book#454
